index.js

```diff
@@ -313,4 +380,116 @@ function reducer(acc, cur) {
   return acc;
 }
 
+/**
+ * Return the response body of an error from a got call if available on the given error object
+ *
+ * @param {Object} error Error object from got call
+ * @return {Response}
+ * @private
+ **/
+const responseFromError = (error) => {
+  if ('response' in error) {
+    const response = error.response;
+    if (response && 'body' in response) {
+      return response;
+    }
+  }
+  return error;
+};
+
+/**
+ * Return true if an error is considered a network level connection error where we weren't actually able to establish contact with Shopify.
+ *
+ * @param {Object} error Error object from got call
+ * @return {Boolean}
+ * @private
+ **/
+const isRetryableConnectionError = (error) => {
+  return (
+    typeof error === 'object' &&
+    'code' in error &&
+    RetryableErrorCodes.has(error.code)
+  );
+};
+
+/**
+ * Given an error from got, see if Shopify told us how long to wait before retrying. Return a duration in ms if we can, and otherwise return null.
+ *
+ * @param {Object} error Error object from got call
+ * @return {Boolean | null}
+ * @private
+ **/
+function maybeRetryMS(error) {
+  // for simplicity, retry network connectivity issues after a hardcoded 1s
+  if (isRetryableConnectionError(error)) {
+    return 1000;
+  }
+
+  const response = responseFromError(error);
+  if (!response) {
+    return null;
+  }
+
+  if (response.headers && response.headers['retry-after']) {
+    const value = parseFloat(response.headers['retry-after']);
+    if (isNaN(value)) {
+      const when = new Date(value).valueOf();
+      return when - Date.now().valueOf();
+    } else if (isFinite(value)) {
+      return value;
+    } else {
+      return null;
+    }
+  }
+
+  if (
+    response.statusCode == 429 ||
+    (response.statusCode >= 500 && response.statusCode < 600)
+  ) {
+    // Arbitrary 2 seconds, in case we get a 429 without a Retry-After response header, or some 5xx response
+    return 2 * 1000;
+  }
+
+  // detect graphql request throttling
+  if (response.body && typeof response.body === 'object') {
+    const body = response.body;
+
+    if (
+      body.errors &&
+      Array.isArray(body.errors) &&
```

Comment:
Isn't `body.errors` always an array?

Reply:
It is but since it is coming from the outside world I think it is prudent to not assume the exact structure, especially when handling errors?

Reply:
Something similar is done in `Shopify.prototype.graphql()` and we are skipping this and the following (`typeof body.errors[0] === 'object'`) check.